### PR TITLE
[Cloud] user settings security page

### PR DESCRIPTION
**Pull request by artemruts**
### Description

Fixes part of #17166 (External Accounts sign-ins)

Adds code external accounts sign-in (GitHub, GitLab) _while trying not to touch much of the original password forms_ (those need to be migrated to hooks, func components).

<details>
  <summary>Outdated backend changes ⌛️</summary>

### Backend changes

I had to add `isPasswordSet` to the `User` type and expose it through GraphQL, here's why I think we'll need it or something similar. 

User:
1. creates an account using BasicAuth (login and password)
2. adds GitHub External Account to sign-in with
3. navigates to `Account security` (`/users/this-user/settings/security`) settings

Since the user has External Account - we're not going to show the `password` form.

4. removes GitHub External Account
5. navigates back to `Account security` (`/users/this-user/settings/security`) settings

Now the user doesn't have any External Account we're going to show the `password` form.  
We don't know which one. If the user previously used BasicAuth, we should show a reset form (with an old password field), otherwise, it's a create password form (without an old password field).

A new `isPasswordSet` field should solve that because we'll know if the user has a valid, not expired password or not  

</details>

### ⚠️ Notes 

Decided to keep previous version of the `Passwords` page while we test the new Cloud GA pages together.  
Right now there's a bit pf duplication but that will go away as soon as I'm confident in the new `Account security` page.  
`Account security` page will be toggled off by default.

### Demo

**When BasicAuth was used to sign-up**

https://user-images.githubusercontent.com/1319181/106505545-770aca80-6496-11eb-8b32-48b53e741dca.mov  
  

**When user doesn't have a password**

Same thing but user's don't see the `old password` field and we send a different GraphQL mutation

When password is not set

### CODE REVIEW

**Comment by quinnkeast** [comment]
> confirm what auth providers cloud will have, can Cloud have multiple GitHub providers, etc?

For this milestone, Cloud users can only have one code host connection for each provider, and we'll only support GitHub and GitLab.

File: internal/types/types.go [hunk 1]
```diff
@@ -965,6 +965,7 @@ type User struct {
 	UpdatedAt             time.Time
 	SiteAdmin             bool
 	BuiltinAuth           bool
+	IsPasswordSet         bool
```
Comment (ryanslade) [inline internal/types/types.go:968]: I don't think this is used anywhere

**Review by quinnkeast — APPROVED** [review]

File: internal/database/users.go [hunk 2]
```diff
@@ -856,11 +856,15 @@ func (u *UserStore) getBySQL(ctx context.Context, query *sqlf.Query) ([]*types.U
 	return users, nil
 }
 
-func (u *UserStore) IsPassword(ctx context.Context, id int32, password string) (bool, error) {
+func (u *UserStore) getPassword(ctx context.Context, id int32, passwd *sql.NullString) error {
```
Comment (ryanslade) [inline internal/database/users.go:859]: nit: I'd personally prefer to have this function return (sql.NullString, error). 

Passing values around to be mutated is generally only used when it's really necessary. It does make the code a little more verbose but I think it's less error prone. For example, what happens if someone passed in a nil passwd by mistake ?

File: internal/database/users.go [hunk 3]
```diff
@@ -869,6 +873,20 @@ func (u *UserStore) IsPassword(ctx context.Context, id int32, password string) (
...
+
+	if err := u.getPassword(ctx, id, &passwd); err != nil {
+		return false, err
+	}
+
+	if passwd.Valid {
+		return true, nil
+	}
+
+	return false, nil
```
Comment (ryanslade) [inline internal/database/users.go:887]:
```suggestion
	return passwd.Valid, nil
```

**Review by ryanslade — APPROVED** [review]
Backend code LGTM

**Comment by ryanslade** [comment]
@artemruts I just realised that we have `user.BuiltinAuth` which indicates where their password is `NULL`, so I believe we can use that instead of adding the new method.

**Review by ryanslade — CHANGES_REQUESTED** [review]
See my latest comment. I think we can use the `user.BuiltinAuth` to check for a null password

**Comment by artemruts** [comment]
> See my latest comment. I think we can use the `user.BuiltinAuth` to check for a null password

Thanks, @ryanslade! `BuiltinAuth` will work, just tested in Dev. Will cleanup backend changes 🙇

**Comment by artemruts** [comment]
@quinnkeast I'm wondering if we should add these changes (`Account security` page) under the cloud "feature flag" and still keep the old `Passwords` page as it is before GA? That would cause temp code duplication but would be resolved before we go GA. We'd simply remove the `Passwords` page. It seems safer that way.

WDYT?

**Comment by quinnkeast** [comment]
Great question!

> It seems safer that way.

Suppose we ship this, and something goes wrong. What went wrong, and what are the worst things that can happen?

**Comment by artemruts** [comment]
@quinnkeast
>Suppose we ship this, and something goes wrong. What went wrong, and what are the worst things that can happen?

Most likely nothing will blow up 😆
Basic logic is:

We display ExternalAccounts only for GitHub/Gitlab and only if we have Auth providers for them.
Right now only GitHub is configured for DotCom so that's the only one we should see.

I'd be down to merge as is if I can test these changes in more "production" env, our current "feature toggle" setup is not flexible for me to:
- merge as is
- enable only for myself
- test and verify
- enable for all of the users

So I could figure out:
1. how to test in more prod env (canary deploy) 
2. 2-step way
   - add back the `Passwords` page, add "feature toggle" for `Account security` page
   - deploy to prod
   - toggle for myself only, test and verify
   - delete "Passwords" page, remove "feature toggle" for `Account security` page
   - deploy again

**Comment by quinnkeast** [comment]
Great breakdown. 

I feel we should go with #2 for now, with the idea we can enable it for the Sourcegraph team to use for a short while to catch any problems before we enable for Cloud as a whole. Better feature flagging will help us a lot here, but short term I think it's a lower priority.

**Review by ryanslade — APPROVED** [review]
Easy review now as no Go code has changed 😄

**Pull request merged**